### PR TITLE
feat: local auth reroute

### PR DESCRIPTION
**Pull request by Magdaline-A**
# DCMAW-9329: User who deletes their biometrics and passcode sees the Sign in screen

Currently, a local auth user who deletes their biometrics and passcode (from device settings) is stranded on the Splash page when they return to the app.

This ticket implements specific logic to ensure this does not occur.

# Checklist

## Before raising your pull request:
- [x] Commit messages that conform to conventional commit messages
- [x] Ran the app locally ensuring it builds 
- [x] Ran the tests locally ensuring they pass on Build
- [x] Pull request has a clear title with a short description about the feature or update
- [x] Created a `draft` pull request if it is not yet ready for review

## Before your pull request can be reviewed:
- [x] Met all of the acceptance criteria specified in the user story on Jira
- [x] Reviewed your own code to ensure you are following the style guidelines
- [x] Ran the app and tested the feature on a range of device sizes
      Please include iPod Touch, iPhone SE and iPhone 11 as a minimum.
- [x] Written Unit and Integration tests if needed

~- [ ] Met all accessibility requirements?~
    ~- [ ] Checked dynamic type sizes are applied~
    ~- [ ] Checked VoiceOver can navigate your new code~
    ~- [ ] Checked a user can navigate only using a keyboard around your new code~ 

## Before merging your pull request:
- [x] Ensure that the code coverage and SonarCloud checks have passed
- [x] Actioned and resolved all comments, reaching out to reviewers for clarifications if necessary.
- [x] Ran the app to ensure that no regressions have been caused by changes during code review.

### CODE REVIEW

File: OneLogin.xcodeproj/project.pbxproj [hunk 1]
```diff
@@ -2000,7 +2000,7 @@
 				ASSETCATALOG_COMPILER_GLOBAL_ACCENT_COLOR_NAME = AccentColor;
 				BASE_URL = mobile.staging.account.gov.uk;
 				CODE_SIGN_STYLE = Automatic;
-				CURRENT_PROJECT_VERSION = 2;
+				CURRENT_PROJECT_VERSION = 1;
```
Comment (obinns-dd) [inline OneLogin.xcodeproj/project.pbxproj:2003]: what's the reason for this change? 👀 
doesn't seem like it's part of the ticket 🙂

Comment (Magdaline-A) [reply OneLogin.xcodeproj/project.pbxproj:2003]: Just forgot to revert it 😅. It was just so I could get around not having a credentialIssuer before your PR merged. But will change it back now!

**Review by benmckillop — APPROVED** [review]

File: Sources/Login/LoginCoordinator.swift [hunk 2]
```diff
@@ -53,6 +54,17 @@ final class LoginCoordinator: NSObject,
...
+        if loginError as? LocalAuthenticationError == .noBiometricsOrPasscode {
+            let signOutWarningScreen = ErrorPresenter
+                .createSignOutWarning(analyticsService: analyticsCenter.analyticsService) {
+                    self.root.dismiss(animated: true)
+                }
+            signOutWarningScreen.modalPresentationStyle = .overFullScreen
+            root.present(signOutWarningScreen, animated: false)
+        }
+    }
+
```
[the remark this answers is not in the document]
Comment (Magdaline-A) [reply Sources/Login/LoginCoordinator.swift:67]: The user being directed to the splash screen is the current implementation. From my understanding, the requirements were for the user to be directed to the 'You’ve been signed out' page and then to the 'sign in' page once the button is tapped

[the remark this answers is not in the document]
Comment (dylanjames-dd) [reply Sources/Login/LoginCoordinator.swift:67]: Oh yes, my mistake 🙂

File: Sources/Application/MainCoordinator.swift [hunk 3]
```diff
@@ -64,6 +64,11 @@ final class MainCoordinator: NSObject,
             fullLogin(loginError: TokenError.expired)
             return
         }
+        
+        guard sessionManager.localAuthentication.canUseLocalAuth(type: .deviceOwnerAuthentication) else {
+            fullLogin(loginError: LocalAuthenticationError.noBiometricsOrPasscode)
+            return
+        }
```
[the remark this answers is not in the document]
Comment (dylanjames-dd) [reply Sources/Application/MainCoordinator.swift:71]: N.B @dylanjames-dd
Need to test what error returned from Secure Store when removing local auth from that set in the instance when saving tokens

**Review by dylanjames-dd — APPROVED** [review]

File: Sources/Login/LoginCoordinator.swift [hunk 4]
```diff
@@ -53,6 +54,17 @@ final class LoginCoordinator: NSObject,
                          object: nil)
     }
     
+    private func checkLocalAuth() {
+        if loginError as? LocalAuthenticationError == .noBiometricsOrPasscode {
```
Comment (obinns-dd) [inline Sources/Login/LoginCoordinator.swift:58]: This might read better as a `guard` statement 🙂

File: Sources/Utilities/LocalAuth/LALocalAuthenticationManager.swift [hunk 5]
```diff
@@ -8,6 +8,10 @@ enum LocalAuthenticationType {
     case none
 }
 
+enum LocalAuthenticationError: Error {
```
[the remark this answers is not in the document]
Comment (Magdaline-A) [reply Sources/Utilities/LocalAuth/LALocalAuthenticationManager.swift:11]: Oh.. what _is_ the error thrown by?

[the remark this answers is not in the document]
Comment (obinns-dd) [reply Sources/Utilities/LocalAuth/LALocalAuthenticationManager.swift:11]: It doesn't look like we throw it at all 👀 can pick this up on the call 🙂

File: Sources/Utilities/Storage/Session/PersistentSessionManager.swift [hunk 6]
```diff
@@ -78,7 +78,11 @@ final class PersistentSessionManager: SessionManager {
     var isPersistentSessionIDMissing: Bool {
         persistentID == nil && isReturningUser
     }
-
+    
+    var hasNotRemovedLocalAuth: Bool {
```
Comment (obinns-dd) [inline Sources/Utilities/Storage/Session/PersistentSessionManager.swift:82]: I think we can make this `private` now too 🙂

Comment (Magdaline-A) [reply Sources/Utilities/Storage/Session/PersistentSessionManager.swift:87]: It doesn't let me.. I think because of the protocol?

Comment (obinns-dd) [reply Sources/Utilities/Storage/Session/PersistentSessionManager.swift:87]: Do we need it in the protocol still? I think it's not needed there either since we removed it from the coordinator 🙂

File: Sources/Utilities/Storage/Session/PersistentSessionManager.swift [hunk 7]
```diff
@@ -139,6 +148,18 @@ final class PersistentSessionManager: SessionManager {
...
+        guard expiryDate != nil else {
+            throw PersistentSessionError.noSessionExists
+        }
+        
+        guard isSessionValid else {
+            throw TokenError.expired
+        }
+        
+        guard hasNotRemovedLocalAuth else {
+            throw PersistentSessionError.userRemovedLocalAuth
```
Comment (obinns-dd) [inline Sources/Utilities/Storage/Session/PersistentSessionManager.swift:160]: Please can we add a test that throws here? (and the `TokenError` above).
Would be good to get the coverage showing as high as possible 🙂

**Review by obinns-dd — APPROVED** [review]
Really awesome work - more than happy for this to merge!!
Just make sure we've covered all the pull request checklist requirements 🙂
